possel/model.py

```diff
@@ -0,0 +1,654 @@
+
+
+class ServerAlreadyAttachedError(Error):
+    """ Exception thrown when someone tries to attach a server handler to a server interface that already has one. """
+
+
+class KeyDefaultDict(collections.defaultdict):
+    """ defaultdict modification that provides the key to the factory. """
+    def __missing__(self, key):
+        if self.default_factory is not None:
+            self[key] = self.default_factory(key)
+            return self[key]
+        else:
+            return super(KeyDefaultDict, self).__missing__(key)
+
+
+database = p.Proxy()
+
+
+class BaseModel(p.Model):
+    class Meta:
+        database = database
+
+    def to_dict(self):
+        return shortcuts.model_to_dict(self, recurse=False)
+
+
+class UserDetails(BaseModel):
+    """ Models all details that are pertinent to every user.
+
+    Users that are connected to a server should use IRCUserModel, this is only really for defining our own details (e.g.
+    so we can send the "NICK ..." message at initial connection).
+    """
+    nick = p.CharField()
+    realname = p.TextField(null=True)
+    username = p.TextField(null=True)
+
+
+class IRCServerModel(BaseModel):
+    """ Models an IRC server, including how to connect to it.
+
+    Does *not* reference networks yet, if ever.
+    """
+    # =========================================================================
+    # Connection Details
+    # ------------------
+    #
+    # Standard IRC connection stuff.
+    #
+    # Doesn't interact with protocol objects; interesting to note?
+    # =========================================================================
+    host = p.TextField()
+    port = p.IntegerField(default=6697)  # You're damn right we default to SSL
+    secure = p.BooleanField(default=True)
+    # =========================================================================
+
+    # =========================================================================
+    # User Details
+    # ------------
+    #
+    # Who *we* are on this server.
+    # =========================================================================
+    user = p.ForeignKeyField(UserDetails)
+    # =========================================================================
+
+    class Meta:
+        indexes = ((('host', 'port'), True),
+                   )
+
+
+class IRCUserModel(UserDetails):
+    """ Models users that are connected to an IRC Server.
+
+    Inherits most of its fields from UserDetails.
+
+    Don't delete user models, they're needed for line models.
+    """
+    host = p.TextField(null=True)  # where they're coming from
+    server = p.ForeignKeyField(IRCServerModel, related_name='users', on_delete='CASCADE')
+    current = p.BooleanField()  # are they connected?
+
+    class Meta:
+        indexes = ((('server', 'nick'), False),
+                   (('server', 'nick', 'current'), False),
+                   )
+
+
+buffer_types = [('normal', 'Standard Buffer'),
```

Comment:
Yes. Last time I tried this my linter didn't like it but it's fine with it now. !resolve